File: src/leetcode/python/ReverseLinkedList_206.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class ReverseLinkedList_206(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return None
        elif not head.next:
            return head

        # tail = head.next
        # head.next = None
        tail, head.next = head.next, None
        while tail:
            tmp = tail.next
            # tail.next = head
            # head = tail
            # tail = tmp
            tail.next, head, tail = head, tail, tmp
        return head

File: src/leetcode/java/ReverseLinkedList_206.java
package leetcode.java.easy;

public class ReverseLinkedList_206 {
  public ListNode reverseList(ListNode head) {
    if (head == null) return null;
    else if (head.next == null) return head;

    ListNode tmp = null;
    ListNode next = head.next;
    head.next = null;
    while (next != null) {
      tmp = next.next;
      next.next = head;
      head = next;
      next = tmp;
    }
    return head;
  }

  class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
      val = x;
      next = null;
    }
  }
}
